### DISCUSSION TITLE
Migrate from new_array_type/register_type?

### DISCUSSION THREAD

**Discussion by di** [psycopg/psycopg#229] [Q&A]
I'm trying to migrate from using `psycopg2.extensions.new_array_type` and `psycopg2.extensions.register_type` in psycopg2 to psycopg3 and it's not clear to me how to do it.

I'm aware of [`psysopg.types.TypesRegistry`](https://www.psycopg.org/psycopg3/docs/api/types.html#psycopg.types.TypesRegistry) but this doesn't seem to provide the same functionality from what I can tell.

**Comment by di** [comment]
Seems like this could be [`psycopg.types.array.register_array`](https://github.com/psycopg/psycopg/blob/88643297dd0766751aa15fdbbd6459a52da13cfa/psycopg/psycopg/types/array.py#L415) but I don't see that documented as a public API.

**Comment by dvarrazzo** [comment]
[`TypeInfo.register()`](https://www.psycopg.org/psycopg3/docs/api/types.html#psycopg.types.TypeInfo.register) should add the array information. If you do that and then register your adapters to the base type, the array should work automatically.

Happy to assist, let me know how it goes on.

**Comment by di** [comment]
Sorry, I might need a little more guidance, I'm not very familiar with psycopg. 🙂 

For context, this is for https://github.com/mahmoudimus/sqlalchemy-citext/issues/26, which is [using `psycopg2` like this](https://github.com/mahmoudimus/sqlalchemy-citext/blob/ac2215c2543958ecabcc172c495d132d4efd8d7a/citext/__init__.py#L48-L49):

```python
array_type = psycopg2.extensions.new_array_type(oids, 'citext[]', psycopg2.STRING)
psycopg2.extensions.register_type(array_type, None)
```

**Comment by dvarrazzo** [comment]
You can look at [`register_hstore()`](https://github.com/psycopg/psycopg/blob/88643297dd0766751aa15fdbbd6459a52da13cfa/psycopg/psycopg/types/hstore.py#L99) to get an idea of the operations involved. I would expect these as the steps to follow to adapt a new type from scratch.

> Sorry, I might need a little more guidance, I'm not very familiar with psycopg. :slightly_smiling_face:

I guess nobody is, it's too new :smile: 

> citext

So, citext is - as far as the client is concerned - text. In what you pasted above, psycopg2 uses the same loader used for the string. In psycopg 3 we can do the same.

The bits of information we need are:

1) what is the citext oid and its array oid?
2) how do convert from Postgres citext to Python string

the bits we don't need are:

3) how do we convert from Postgres citext array to Python list of strings (because that's done automatically)
4) how do we convert from Python `str` to PostgreSQL citext (because just converting to string should do)

So the steps are:

1) we use `TypeInfo.fetch()` to find the citext oid (it's different in any db because it's an extension):

```python
cnn.execute("create extension citext")
citext = psycopg.types.TypeInfo.fetch(cnn, "citext")
print(citext)
<TypeInfo: citext (oid: 9319139, array oid: 9319144)>  # the oids depend on the database

# You can register the same object on other connections as long as they are on the same database,
# or globally as long as your program uses only one db
citext.register(cnn)
```

2) Now we should register a loader to convert citext -> str, and because citext looks pretty much like text we can use the text loader:

```python
cnn.adapters.register_loader("citext", psycopg.types.string.TextLoader)

# this too can be global if `psycopg.adapters` is used instead.
```

Now the connection will know of citext and of citext array:

```python
>>> cnn.execute("select 'hello'::citext").fetchone()[0]
'hello'
>>> cnn.execute("select '{hello,there}'::citext[]").fetchone()[0]
['hello', 'there']
```

TBH probably citext doesn't need the TextLoader registration, just because the loader for the unknown types does the same operation. So step 2) might be optional here, but not for other data types.

Does this help?
